services/worker-manager/src/providers/azure.js

```diff
@@ -451,49 +671,158 @@ class AzureProvider extends Provider {
+    }
+
+    // NB: possible resource leak if we don't require `return true`
+    // we don't check operation status: no differentiating between
+    // operation => create and operation => delete
+    if (typeData.id || shouldDelete) {
+      // we need to delete the resource
+      let deleteRequest = await this._enqueue('query', () => client.beginDeleteMethod(
+        worker.providerData.resourceGroupName,
+        typeData.name,
+      ));
+      // track operation
+      await worker.modify(w => {
+        w.providerData[resourceType].id = false;
+        w.providerData[resourceType].operation = deleteRequest.getPollState().azureAsyncOperationHeaderValue;
+      });
+    }
+    return false;
   }
 
+  /*
+   * removeWorker marks a worker for deletion and begins removal
+   *
+   * worker: the worker for which the resource is being provisioned
+   * client: the Azure SDK client for the resource
+   * resourceType: the short name used to identify the resource in providerData
+   */
   async removeWorker({worker}) {
-    await this._removeWorker({...worker.providerData});
+    let states = this.Worker.states;
+    if (worker.state === states.STOPPED) {
+      // we're done
+      return states.STOPPED;
+    }
+
+    let state = states.STOPPING;
+    // After we make the delete request we set id to false
+    // some delete operations (i.e. VMs) take a long time though
+    // we use the result of removeResource to _ensure_ deletion has completed
+    // before moving on to the next step, so that we don't leak resources
+    try {
+      // VM must be deleted before disk
+      // VM must be deleted before NIC
+      // NIC must be deleted before IP
+      let vmDeleted = await this.removeResource({
+        worker,
+        client: this.computeClient.virtualMachines,
+        resourceType: 'vm',
+      });
+      if (!vmDeleted || worker.providerData.vm.id) {
+        return state;
+      }
+      let nicDeleted = await this.removeResource({
+        worker,
+        client: this.networkClient.networkInterfaces,
+        resourceType: 'nic',
+      });
+      if (!nicDeleted || worker.providerData.nic.id) {
+        return state;
+      }
+      let ipDeleted = await this.removeResource({
+        worker,
+        client: this.networkClient.publicIPAddresses,
+        resourceType: 'ip',
+      });
+      if (!ipDeleted || worker.providerData.ip.id) {
+        return state;
+      }
+      let diskDeleted = await this.removeResource({
+        worker,
+        client: this.computeClient.disks,
+        resourceType: 'disk',
+      });
+      if (!diskDeleted || worker.providerData.disk.id) {
+        return state;
+      }
+      // change to stopped
+      state = states.STOPPED;
+      await worker.modify(w => {
+        const now = new Date();
+        w.lastModified = now;
+        w.lastChecked = now;
+        w.state = state;
+      });
+    } catch (err) {
+      // if this is called directly and not via checkWorker may not exist
```

Comment:
.. is this the case when this is called by the REST API?  In that case, what happens to the errors?

Reply:
The only case where this has been necessary that I've seen is in the tests. I haven't tested the REST API for deleteWorker() - this comment creates those things, i.e. handles the case where they don't exist.